### ISSUE TITLE
Types break in Typescript 4.8

### ISSUE THREAD

**Issue by okomarov** [stalniy/casl#673]
**Describe the bug**
Types in `conditions.d.ts` break with Typescript 4.8 due to [these changes](https://devblogs.microsoft.com/typescript/announcing-typescript-4-8-rc/#improved-intersection-reduction-union-compatibility-and-narrowing).

**To Reproduce**
https://replit.com/@OlegAzava/CASL-and-typescript-48#index.ts

In the console, run `yarn build`:

<details><summary>Error details</summary>
<p>

```ts
node_modules/@casl/ability/dist/types/matchers/conditions.d.ts:28:76 - error TS2344: Type 'T' does not satisfy the constraint '{}'.

28 export declare type MongoQuery<T = AnyObject> = BuildMongoQuery<MergeUnion<T>, {
                                                                              ~

  node_modules/@casl/ability/dist/types/matchers/conditions.d.ts:28:32
    28 export declare type MongoQuery<T = AnyObject> = BuildMongoQuery<MergeUnion<T>, {
                                      ~~~~~~~~~~~~~
    This type parameter might need an `extends {}` constraint.

node_modules/@casl/ability/dist/types/matchers/conditions.d.ts:30:45 - error TS2344: Type 'T' does not satisfy the constraint '{}'.

30     field: Pick<DefaultOperators<MergeUnion<T>>['field'], keyof typeof defaultInstructions>;
                                               ~

  node_modules/@casl/ability/dist/types/matchers/conditions.d.ts:28:32
    28 export declare type MongoQuery<T = AnyObject> = BuildMongoQuery<MergeUnion<T>, {
                                      ~~~~~~~~~~~~~
    This type parameter might need an `extends {}` constraint.


Found 2 errors in the same file, starting at: node_modules/@casl/ability/dist/types/matchers/conditions.d.ts:28
```

</p>
</details>

**Expected behavior**
No type errors.

**CASL Version**
@casl/ability 6.1.1

**Environment**:
Node 16.14.0

**Comment by okomarov** [comment]
@stalniy What's the best way to help with this without having to install pnpm? Is it possible to reproduce the setup in the provided replit so other could eventually fork it and provide submissions off that?

**Comment by stalniy** [comment]
You can install https://volta.sh/ very handy in managing node and node tools versions (like yarn, pnpm,etc). You need to install pnpm

**Comment by stalniy** [comment]
I'll fix it today, thanks for the report

**Comment by stalniy** [comment]
fixed in @casl/ability@6.1.2
